File: agents/AGT-CLAUDE-CAPTURE-1__0368F157/main.py
#!/usr/bin/env python3
"""
AGT-CLAUDE-CAPTURE-1: Claude Code Session Capture Agent

Captures Claude Code sessions and streams them to Redis for consumption
by the conversation streaming pipeline. Designed as a development tool
to capture architectural discussions that would otherwise be lost.

Usage:
    python main.py [--mode=hook|standalone]
"""

import asyncio
import json
import logging
import os
import sys
import time
import uuid
from datetime import datetime, timezone
from typing import Dict, Any, Optional, List
from dataclasses import dataclass, asdict
from pathlib import Path

import redis
import psutil
import aiofiles
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler

@dataclass
class ConversationTurn:
    """Single conversation turn (user input + assistant response)"""
    session_id: str
    agent_id: str
    timestamp: str
    turn_count: int
    user: str
    assistant: str

@dataclass
class SessionEvent:
    """Session lifecycle event"""
    event_type: str  # start, message, heartbeat, end
    session_id: str
    timestamp: str
    data: Optional[Dict[str, Any]] = None

class ClaudeCaptureAgent:
    """Claude Code session capture and streaming agent"""
    
    def __init__(self):
        self.agent_id = "AGT-CLAUDE-CAPTURE-1"
        self.cid = "cid:centerfire:agent:0368F157" 
        self.redis_client = None
        self.active_sessions: Dict[str, Dict] = {}
        self.conversation_buffer: List[str] = []
        self.session_id = None
        
        # Lazy disk buffering for Redis failures
        self.redis_healthy = True
        self.disk_buffer_active = False
        self.buffer_file_path = Path("./redis_failover_buffer.jsonl")
        self.last_redis_check = 0
        self.redis_check_interval = 30  # Check Redis health every 30 seconds
        
        # Redis channels and streams
        self.request_channel = "agent.claude-capture.request"
        self.response_channel = "agent.claude-capture.response"
        self.conversation_stream = "centerfire:semantic:conversations"
        self.session_stream = "claude:sessions:stream"
        
        # Setup logging
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        )
        self.logger = logging.getLogger(self.agent_id)
        
    async def start(self):
        """Start the Claude capture agent"""
        self.logger.info(f"🚀 Starting {self.agent_id}...")
        
        # Connect to Redis
        await self._connect_redis()
        
        # Register with AGT-MANAGER-1
        await self._register_with_manager()
        
        # Start background tasks
        tasks = [
            self._listen_for_requests(),
            self._monitor_claude_processes(),
            self._capture_current_session()
        ]
        
        self.logger.info(f"✅ {self.agent_id} ready for Claude Code session capture")
        
        # Run all tasks concurrently
        try:
            await asyncio.gather(*tasks)
        except KeyboardInterrupt:
            self.logger.info(f"🛑 {self.agent_id} shutting down...")
            await self._unregister_with_manager()
    
    async def _connect_redis(self):
        """Connect to Redis"""
        try:
            self.redis_client = redis.Redis(
                host="localhost", 
                port=6380, 
                decode_responses=True
            )
            # Test connection
            self.redis_client.ping()
            self.logger.info("📡 Connected to Redis successfully")
        except Exception as e:
            self.logger.error(f"❌ Failed to connect to Redis: {e}")
            raise
    
    async def _register_with_manager(self):
        """Register with AGT-MANAGER-1"""
        session_data = {
            "pid": os.getpid(),
            "agent_type": "persistent",
            "capabilities": ["session_capture", "conversation_streaming", "claude_monitoring"],
            "channels": [self.request_channel],
            "language": "python"
        }
        
        registration_data = {
            "request_type": "register_running",
            "agent_name": self.agent_id,
            "session_data": session_data
        }
        
        try:
            self.redis_client.publish("centerfire:agent:manager", json.dumps(registration_data))
            self.logger.info(f"📝 Registered with AGT-MANAGER-1 (PID: {os.getpid()})")
            
            # Start heartbeat
            await self._start_heartbeat()
        except Exception as e:
            self.logger.error(f"❌ Failed to register with manager: {e}")
    
    async def _start_heartbeat(self):
        """Start periodic heartbeat to AGT-MANAGER-1"""
        async def heartbeat_loop():
            while True:
                try:
                    await asyncio.sleep(30)  # Heartbeat every 30 seconds
                    
                    heartbeat_data = {
                        "request_type": "heartbeat",
                        "agent_name": self.agent_id
                    }
                    
                    self.redis_client.publish("centerfire:agent:manager", json.dumps(heartbeat_data))
                    
                except Exception as e:
                    self.logger.error(f"❌ Heartbeat error: {e}")
                    await asyncio.sleep(60)
        
        # Start heartbeat in the background
        asyncio.create_task(heartbeat_loop())
    
    async def _unregister_with_manager(self):
        """Unregister from AGT-MANAGER-1"""
        unregister_data = {
            "request_type": "unregister_running",
            "agent_name": self.agent_id
        }
        
        try:
            self.redis_client.publish("centerfire:agent:manager", json.dumps(unregister_data))
            self.logger.info("📋 Unregistered from AGT-MANAGER-1")
        except Exception as e:
            self.logger.error(f"❌ Failed to unregister: {e}")
    
    async def _listen_for_requests(self):
        """Listen for Redis pub/sub requests"""
        pubsub = self.redis_client.pubsub()
        pubsub.subscribe(self.request_channel)
        
        try:
            for message in pubsub.listen():
                if message['type'] == 'message':
                    await self._handle_request(message['data'])
        except Exception as e:
            self.logger.error(f"❌ Error in request listener: {e}")
        finally:
            pubsub.close()
    
    async def _handle_request(self, payload: str):
        """Handle incoming Redis requests"""
        try:
            request = json.loads(payload)
            action = request.get('action')
            request_id = request.get('request_id', str(uuid.uuid4()))
            
            self.logger.info(f"📥 Processing request: {action}")
            
            response = {"request_id": request_id, "success": False}
            
            if action == "capture_session":
                response = await self._handle_capture_session(request)
            elif action == "stream_conversation":  
                response = await self._handle_stream_conversation(request)
            elif action == "get_status":
                response = await self._handle_get_status(request)
            else:
                response["error"] = f"Unknown action: {action}"
            
            # Publish response
            self.redis_client.publish(self.response_channel, json.dumps(response))
            
        except Exception as e:
            self.logger.error(f"❌ Error handling request: {e}")
    
    async def _capture_current_session(self):
        """Capture the current Claude Code session"""
        # This is the critical function - capture THIS conversation
        self.session_id = f"CLAUDE-CODE-SESSION_{int(time.time())}"
        
        self.logger.info(f"🎯 Starting capture of current Claude Code session: {self.session_id}")
        
        # Stream session start event
        await self._stream_session_event("session_started", {
            "session_type": "claude_code", 
            "capture_method": "development_agent",
            "note": "Capturing architectural discussion session"
        })
        
        # For now, we'll capture this conversation by monitoring the process
        # In production, this would hook into Claude Code's actual conversation flow
        await self._monitor_conversation_flow()
    
    async def _monitor_conversation_flow(self):
        """Monitor and capture conversation flow"""
        conversation_data = {
            "session_id": self.session_id,
            "agent_id": "CLAUDE-CODE",  # Since we're Claude Code
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "turn_count": 1,
            "user": "Human architectural discussion about Redis streams, Weaviate context, and semantic ticketing",
            "assistant": "Claude Code implementing conversation streaming pipeline with AGT-CONTEXT-1 integration"
        }
        
        # Stream this conversation
        await self._stream_conversation(conversation_data)
        
        # Continue monitoring (this is a simplified version)
        while True:
            await asyncio.sleep(60)  # Check every minute
            await self._stream_session_event("heartbeat", {
                "active_conversations": len(self.active_sessions),
                "buffer_size": len(self.conversation_buffer)
            })
    
    async def _stream_conversation(self, conversation_data: Dict[str, Any]):
        """Stream conversation data to Redis with lazy disk buffering failover"""
        stream_data = {
            "data": json.dumps(conversation_data)
        }
        
        # Check Redis health periodically
        await self._check_redis_health()
        
        if self.redis_healthy:
            try:
                # Normal path - direct to Redis
                self.redis_client.xadd(self.conversation_stream, stream_data)
                self.logger.info(f"💾 Streamed conversation to {self.conversation_stream}")
                
                # If we were in failover mode, try to replay buffer
                if self.disk_buffer_active:
                    await self._replay_disk_buffer()
                
            except Exception as e:
                self.logger.warning(f"⚠️ Redis write failed, activating disk buffer: {e}")
                self.redis_healthy = False
                await self._activate_disk_buffer()
                await self._write_to_disk(stream_data)
        else:
            # Failover path - disk buffer
            self.logger.debug(f"📝 Writing to disk buffer (Redis unhealthy)")
            await self._write_to_disk(stream_data)
    
    async def _stream_session_event(self, event_type: str, data: Optional[Dict] = None):
        """Stream session lifecycle events"""
        event = SessionEvent(
            event_type=event_type,
            session_id=self.session_id,
            timestamp=datetime.now(timezone.utc).isoformat(),
            data=data
        )
        
        try:
            self.redis_client.xadd(self.session_stream, asdict(event))
            self.logger.info(f"📊 Streamed {event_type} event for session {self.session_id}")
        except Exception as e:
            self.logger.error(f"❌ Failed to stream session event: {e}")
    
    async def _monitor_claude_processes(self):
        """Monitor Claude Code processes"""
        while True:
            try:
                # Look for Claude Code processes
                claude_processes = []
                for proc in psutil.process_iter(['pid', 'name', 'cmdline']):
                    try:
                        if 'claude' in proc.info['name'].lower() or \
                           any('claude' in arg.lower() for arg in (proc.info['cmdline'] or [])):
                            claude_processes.append(proc.info)
                    except (psutil.NoSuchProcess, psutil.AccessDenied):
                        continue
                
                if claude_processes:
                    self.logger.debug(f"🔍 Found {len(claude_processes)} Claude-related processes")
                
                await asyncio.sleep(30)  # Check every 30 seconds
                
            except Exception as e:
                self.logger.error(f"❌ Error monitoring processes: {e}")
                await asyncio.sleep(60)
    
    
    async def _handle_capture_session(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Handle session capture request"""
        session_id = request.get('session_id', f"session_{int(time.time())}")
        
        self.active_sessions[session_id] = {
            "started": time.time(),
            "status": "active",
            "turns": 0
        }
        
        await self._stream_session_event("session_started", {"session_id": session_id})
        
        return {
            "success": True,
            "session_id": session_id,
            "message": "Session capture started"
        }
    
    async def _handle_stream_conversation(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Handle conversation streaming request"""
        conversation_data = request.get('conversation_data', {})
        
        if not conversation_data:
            return {"success": False, "error": "No conversation data provided"}
        
        await self._stream_conversation(conversation_data)
        
        return {
            "success": True,
            "message": "Conversation streamed successfully"
        }
    
    async def _handle_get_status(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Handle status request"""
        return {
            "success": True,
            "agent_id": self.agent_id,
            "cid": self.cid,
            "active_sessions": len(self.active_sessions),
            "buffer_size": len(self.conversation_buffer),
            "current_session": self.session_id,
            "uptime": time.time() - (hasattr(self, '_start_time') and self._start_time or time.time())
        }
    
    async def _check_redis_health(self):
        """Check Redis health periodically"""
        current_time = time.time()
        if current_time - self.last_redis_check < self.redis_check_interval:
            return
            
        self.last_redis_check = current_time
        
        try:
            # Quick ping to test Redis
            self.redis_client.ping()
            if not self.redis_healthy:
                self.logger.info("✅ Redis recovered - switching back from disk buffer")
                self.redis_healthy = True
                # Replay disk buffer will happen in next write
                
        except Exception as e:
            if self.redis_healthy:
                self.logger.warning(f"⚠️ Redis health check failed: {e}")
                self.redis_healthy = False
    
    async def _activate_disk_buffer(self):
        """Activate disk buffering mode"""
        if not self.disk_buffer_active:
            self.disk_buffer_active = True
            self.logger.info(f"🆘 Activating disk buffer: {self.buffer_file_path}")
    
    async def _write_to_disk(self, stream_data: Dict[str, Any]):
        """Write conversation data to disk buffer"""
        try:
            buffer_entry = {
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "stream": self.conversation_stream,
                "data": stream_data
            }
            
            async with aiofiles.open(self.buffer_file_path, 'a') as f:
                await f.write(json.dumps(buffer_entry) + '\n')
                
        except Exception as e:
            self.logger.error(f"❌ Failed to write to disk buffer: {e}")
    
    async def _replay_disk_buffer(self):
        """Replay buffered conversations to Redis when it recovers"""
        if not self.buffer_file_path.exists():
            return
            
        self.logger.info(f"🔄 Replaying disk buffer to Redis...")
        
        try:
            replayed_count = 0
            
            async with aiofiles.open(self.buffer_file_path, 'r') as f:
                async for line in f:
                    if line.strip():
                        buffer_entry = json.loads(line)
                        stream_data = buffer_entry['data']
                        
                        # Replay to Redis
                        self.redis_client.xadd(self.conversation_stream, stream_data)
                        replayed_count += 1
            
            # Clear the buffer file after successful replay
            self.buffer_file_path.unlink()
            self.disk_buffer_active = False
            
            self.logger.info(f"✅ Replayed {replayed_count} buffered conversations to Redis")
            
        except Exception as e:
            self.logger.error(f"❌ Failed to replay disk buffer: {e}")

async def main():
    """Main entry point"""
    agent = ClaudeCaptureAgent()
    agent._start_time = time.time()
    
    try:
        await agent.start()
    except KeyboardInterrupt:
        print(f"\n🛑 {agent.agent_id} shutting down...")
    except Exception as e:
        print(f"❌ Fatal error: {e}")
        return 1
    
    return 0

if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: agents/AGT-CLAUDE-CAPTURE-1__0368F157/main.go
package main

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/redis/go-redis/v9"
)

// ClaudeCaptureAgent - Agent for capturing Claude Code sessions and streaming to Redis
type ClaudeCaptureAgent struct {
	AgentID         string
	CID            string
	RequestChannel  string
	ResponseChannel string
	RedisClient    *redis.Client
	ctx            context.Context
	cancel         context.CancelFunc
}

// SessionData - Structure for Claude Code session data
type SessionData struct {
	SessionID   string                 `json:"session_id"`
	Timestamp   time.Time              `json:"timestamp"`
	MessageType string                 `json:"message_type"` // "user_input", "assistant_response", "system_info"
	Content     string                 `json:"content"`
	Metadata    map[string]interface{} `json:"metadata"`
}

// ConversationEvent - Structure for conversation events
type ConversationEvent struct {
	EventID     string                 `json:"event_id"`
	SessionID   string                 `json:"session_id"`
	Timestamp   time.Time              `json:"timestamp"`
	EventType   string                 `json:"event_type"` // "session_start", "user_message", "assistant_message", "session_end"
	Data        map[string]interface{} `json:"data"`
	AgentSource string                 `json:"agent_source"`
}

// NewAgent - Create new Claude Code capture agent
func NewAgent() *ClaudeCaptureAgent {
	ctx, cancel := context.WithCancel(context.Background())
	
	// Connect to Redis container on port 6380
	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6380",
		Password: "",
		DB:       0,
	})
	
	return &ClaudeCaptureAgent{
		AgentID:         "AGT-CLAUDE-CAPTURE-1",
		CID:            "cid:centerfire:agent:0368F157",
		RequestChannel:  "claude.sessions.request",
		ResponseChannel: "claude.sessions.response",
		RedisClient:    rdb,
		ctx:            ctx,
		cancel:         cancel,
	}
}

// Start - Start the Claude Code capture agent
func (a *ClaudeCaptureAgent) Start() {
	fmt.Printf("%s starting Claude Code session capture...\n", a.AgentID)
	fmt.Printf("CID: %s\n", a.CID)
	fmt.Printf("Monitoring Claude Code sessions and streaming to Redis\n")
	
	// Test Redis connection
	_, err := a.RedisClient.Ping(a.ctx).Result()
	if err != nil {
		fmt.Printf("Failed to connect to Redis: %v\n", err)
		return
	}
	fmt.Printf("Connected to Redis successfully\n")
	
	// Set up graceful shutdown
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
	
	// Start session capture in goroutine
	go a.startSessionCapture()
	
	// Start Redis request listener in goroutine
	go a.startRedisListener()
	
	fmt.Printf("%s ready - capturing Claude Code sessions\n", a.AgentID)
	
	// Wait for shutdown signal
	<-sigChan
	fmt.Printf("\n%s shutting down...\n", a.AgentID)
	a.cancel()
}

// startSessionCapture - Monitor and capture Claude Code sessions
func (a *ClaudeCaptureAgent) startSessionCapture() {
	fmt.Printf("%s: Starting session capture monitoring\n", a.AgentID)
	
	// Create a session for this capture instance
	sessionID := a.generateSessionID()
	
	// Publish session start event
	a.publishConversationEvent(sessionID, "session_start", map[string]interface{}{
		"agent_id": a.AgentID,
		"cid":      a.CID,
		"started_at": time.Now(),
	})
	
	// Monitor standard input for Claude Code interactions
	go a.monitorStdInput(sessionID)
	
	// Keep the capture running
	for {
		select {
		case <-a.ctx.Done():
			// Publish session end event
			a.publishConversationEvent(sessionID, "session_end", map[string]interface{}{
				"agent_id": a.AgentID,
				"ended_at": time.Now(),
			})
			return
		case <-time.After(30 * time.Second):
			// Periodic heartbeat
			a.publishSessionHeartbeat(sessionID)
		}
	}
}

// monitorStdInput - Monitor standard input for Claude Code interactions
func (a *ClaudeCaptureAgent) monitorStdInput(sessionID string) {
	reader := bufio.NewReader(os.Stdin)
	
	for {
		select {
		case <-a.ctx.Done():
			return
		default:
			// Check if there's input available (non-blocking)
			if a.hasInputAvailable() {
				line, err := reader.ReadString('\n')
				if err != nil && err != io.EOF {
					continue
				}
				
				if strings.TrimSpace(line) != "" {
					// Capture user input
					sessionData := SessionData{
						SessionID:   sessionID,
						Timestamp:   time.Now(),
						MessageType: "user_input",
						Content:     strings.TrimSpace(line),
						Metadata: map[string]interface{}{
							"source":    "claude_code",
							"agent_id":  a.AgentID,
							"capture_method": "stdin_monitor",
						},
					}
					
					a.streamSessionData(sessionData)
					
					// Publish conversation event
					a.publishConversationEvent(sessionID, "user_message", map[string]interface{}{
						"content": strings.TrimSpace(line),
						"length":  len(strings.TrimSpace(line)),
					})
				}
			}
			time.Sleep(100 * time.Millisecond)
		}
	}
}

// hasInputAvailable - Check if input is available (simplified version)
func (a *ClaudeCaptureAgent) hasInputAvailable() bool {
	// This is a simplified check - in a real implementation,
	// you might use more sophisticated methods to detect Claude Code sessions
	info, err := os.Stdin.Stat()
	if err != nil {
		return false
	}
	return (info.Mode() & os.ModeCharDevice) == 0
}

// startRedisListener - Listen for Redis requests
func (a *ClaudeCaptureAgent) startRedisListener() {
	// Subscribe to request channel
	pubsub := a.RedisClient.Subscribe(a.ctx, a.RequestChannel)
	defer pubsub.Close()
	
	// Listen for messages
	ch := pubsub.Channel()
	
	fmt.Printf("%s: Redis listener started\n", a.AgentID)
	
	for {
		select {
		case <-a.ctx.Done():
			fmt.Printf("%s: Redis listener stopping\n", a.AgentID)
			return
		case msg := <-ch:
			a.processRedisMessage(msg.Payload)
		}
	}
}

// processRedisMessage - Process incoming Redis message
func (a *ClaudeCaptureAgent) processRedisMessage(payload string) {
	var request map[string]interface{}
	if err := json.Unmarshal([]byte(payload), &request); err != nil {
		fmt.Printf("%s: Error parsing request: %v\n", a.AgentID, err)
		return
	}
	
	fmt.Printf("%s received Redis request: %v\n", a.AgentID, request["action"])
	
	// Handle the request
	response := a.HandleRequest(request)
	
	// Send response back via Redis
	responseData, _ := json.Marshal(response)
	a.RedisClient.Publish(a.ctx, a.ResponseChannel, responseData)
}

// HandleRequest - Handle incoming requests
func (a *ClaudeCaptureAgent) HandleRequest(request map[string]interface{}) map[string]interface{} {
	action, ok := request["action"].(string)
	if !ok {
		return map[string]interface{}{
			"error": "No action specified",
		}
	}
	
	// Route to appropriate handler
	switch action {
	case "capture_session":
		return a.handleCaptureSession(request)
	case "stream_conversation":
		return a.handleStreamConversation(request)
	case "get_session_status":
		return a.handleGetSessionStatus(request)
	default:
		return map[string]interface{}{
			"error": fmt.Sprintf("Unknown action: %s", action),
		}
	}
}

// handleCaptureSession - Handle session capture requests
func (a *ClaudeCaptureAgent) handleCaptureSession(request map[string]interface{}) map[string]interface{} {
	params, ok := request["params"].(map[string]interface{})
	if !ok {
		return map[string]interface{}{"error": "No params provided"}
	}
	
	sessionID := a.generateSessionID()
	if id, ok := params["session_id"].(string); ok {
		sessionID = id
	}
	
	content, ok := params["content"].(string)
	if !ok {
		return map[string]interface{}{"error": "Content required"}
	}
	
	messageType, ok := params["type"].(string)
	if !ok {
		messageType = "captured_session"
	}
	
	// Create session data
	sessionData := SessionData{
		SessionID:   sessionID,
		Timestamp:   time.Now(),
		MessageType: messageType,
		Content:     content,
		Metadata: map[string]interface{}{
			"source":      "api_request",
			"agent_id":    a.AgentID,
			"request_id":  request["request_id"],
		},
	}
	
	// Stream to Redis
	a.streamSessionData(sessionData)
	
	return map[string]interface{}{
		"success":    true,
		"session_id": sessionID,
		"message":    "Session data captured and streamed",
	}
}

// handleStreamConversation - Handle conversation streaming requests
func (a *ClaudeCaptureAgent) handleStreamConversation(request map[string]interface{}) map[string]interface{} {
	params, ok := request["params"].(map[string]interface{})
	if !ok {
		return map[string]interface{}{"error": "No params provided"}
	}
	
	sessionID, ok := params["session_id"].(string)
	if !ok {
		return map[string]interface{}{"error": "Session ID required"}
	}
	
	eventType, ok := params["event_type"].(string)
	if !ok {
		eventType = "conversation_event"
	}
	
	data, ok := params["data"].(map[string]interface{})
	if !ok {
		data = make(map[string]interface{})
	}
	
	// Publish conversation event
	a.publishConversationEvent(sessionID, eventType, data)
	
	return map[string]interface{}{
		"success":   true,
		"session_id": sessionID,
		"message":   "Conversation event streamed",
	}
}

// handleGetSessionStatus - Handle session status requests
func (a *ClaudeCaptureAgent) handleGetSessionStatus(request map[string]interface{}) map[string]interface{} {
	return map[string]interface{}{
		"agent_id":     a.AgentID,
		"cid":          a.CID,
		"status":       "running",
		"uptime":       time.Since(time.Now()).String(),
		"capabilities": []string{"session_capture", "redis_streaming", "conversation_tracking"},
	}
}

// streamSessionData - Stream session data to Redis
func (a *ClaudeCaptureAgent) streamSessionData(data SessionData) {
	streamName := "claude:sessions:stream"
	
	dataJSON, err := json.Marshal(data)
	if err != nil {
		fmt.Printf("%s: Error marshaling session data: %v\n", a.AgentID, err)
		return
	}
	
	_, err = a.RedisClient.XAdd(a.ctx, &redis.XAddArgs{
		Stream: streamName,
		Values: map[string]interface{}{
			"session_id":   data.SessionID,
			"message_type": data.MessageType,
			"content":      data.Content,
			"data":         string(dataJSON),
			"timestamp":    data.Timestamp.Unix(),
			"source":       a.AgentID,
		},
	}).Result()
	
	if err != nil {
		fmt.Printf("%s: Error streaming session data: %v\n", a.AgentID, err)
	} else {
		fmt.Printf("%s: Streamed session data for session %s\n", a.AgentID, data.SessionID)
	}
}

// publishConversationEvent - Publish conversation events to Redis
func (a *ClaudeCaptureAgent) publishConversationEvent(sessionID, eventType string, data map[string]interface{}) {
	streamName := "claude:conversations:stream"
	
	event := ConversationEvent{
		EventID:     fmt.Sprintf("evt_%d", time.Now().UnixNano()),
		SessionID:   sessionID,
		Timestamp:   time.Now(),
		EventType:   eventType,
		Data:        data,
		AgentSource: a.AgentID,
	}
	
	eventJSON, err := json.Marshal(event)
	if err != nil {
		fmt.Printf("%s: Error marshaling conversation event: %v\n", a.AgentID, err)
		return
	}
	
	_, err = a.RedisClient.XAdd(a.ctx, &redis.XAddArgs{
		Stream: streamName,
		Values: map[string]interface{}{
			"event_id":    event.EventID,
			"session_id":  event.SessionID,
			"event_type":  event.EventType,
			"data":        string(eventJSON),
			"timestamp":   event.Timestamp.Unix(),
			"source":      a.AgentID,
		},
	}).Result()
	
	if err != nil {
		fmt.Printf("%s: Error publishing conversation event: %v\n", a.AgentID, err)
	} else {
		fmt.Printf("%s: Published conversation event: %s for session %s\n", a.AgentID, eventType, sessionID)
	}
}

// publishSessionHeartbeat - Publish periodic heartbeat for active sessions
func (a *ClaudeCaptureAgent) publishSessionHeartbeat(sessionID string) {
	a.publishConversationEvent(sessionID, "session_heartbeat", map[string]interface{}{
		"agent_id": a.AgentID,
		"status":   "active",
		"heartbeat_time": time.Now(),
	})
}

// generateSessionID - Generate unique session ID
func (a *ClaudeCaptureAgent) generateSessionID() string {
	return fmt.Sprintf("claude_session_%d", time.Now().UnixNano())
}

func main() {
	agent := NewAgent()
	agent.Start()
}